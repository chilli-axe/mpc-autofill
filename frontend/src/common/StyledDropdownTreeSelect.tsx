import DropdownTreeSelect from "react-dropdown-tree-select";
import styled from "styled-components";

export const StyledDropdownTreeSelect = styled(DropdownTreeSelect)`
  .tag {
    color: black;
    background-color: #dddddd;
  }
  .tag-remove {
    color: #666666;
  }

  .dropdown-trigger {
    border-radius: 0.25rem;
    background-color: white;
  }
  .dropdown-content {
    border-radius: 0.25rem;
  }

  .search {
    background-color: white;
    color: black;
  }
  .search::placeholder {
    color: black;
  }
  .placeholder {
    color: gray;
  }

  .toggle {
    font: normal normal normal 12px/1 bootstrap-icons;
    top: 2px;
    left: 2px;
  }

  .toggle.collapsed::after {
    content: "\F4FA";
  }

  .toggle.expanded::after {
    content: "\F2E6";
  }

  color: black;

  .root {
    padding: 0;
    margin: 0;
  }
`;
