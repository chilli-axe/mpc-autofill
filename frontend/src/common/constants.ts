import { CardType as CardTypeSchema } from "@/common/schema_types";
import { Cardstock, CardType, Faces, SortBy } from "@/common/types";
export const ProjectName = "MPC Autofill";
export const MakePlayingCards = "MakePlayingCards.com";
export const MakePlayingCardsURL = "https://www.makeplayingcards.com";

export const Card: CardType = CardTypeSchema.Card;
export const Cardback: CardType = CardTypeSchema.Cardback;
export const Token: CardType = CardTypeSchema.Token;

export const SelectedImageSeparator = "@";
export const CardTypeSeparator = ":";
export const FaceSeparator = "//";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";

export const ToggleButtonHeight = 38; // pixels
export const NavbarHeight = 50; // pixels - aligns with the natural height of the navbar
export const RibbonHeight = 54; // pixels
export const NavbarLogoHeight = 40; // pixels
export const ContentMaxWidth = 1200; // pixels - aligns with bootstrap's large breakpoint

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
  SearchResults = "searchResults",
  SampleCards = "sampleCards",
}

export const S27: Cardstock = "(S27) Smooth";
export const S30: Cardstock = "(S30) Standard Smooth";
export const S33: Cardstock = "(S33) Superior Smooth";
export const M31: Cardstock = "(M31) Linen";
export const P10: Cardstock = "(P10) Plastic";
export const Cardstocks: Array<Cardstock> = [S27, S30, S33, M31, P10];

export const CardstockFoilCompatibility: { [cardstock in Cardstock]: boolean } =
  {
    [S27]: true,
    [S30]: true,
    [S33]: true,
    [M31]: true,
    [P10]: false,
  };

export const GoogleDriveImageAPIURL =
  "https://script.google.com/macros/s/AKfycbw8laScKBfxda2Wb0g63gkYDBdy8NWNxINoC4xDOwnCQ3JMFdruam1MdmNmN4wI5k4/exec";

export const SearchResultsEndpointPageSize = 300;
export const CardEndpointPageSize = 1000;

export enum CSVHeaders {
  quantity = "Quantity",
  frontQuery = "Front",
  frontSelectedImage = "Front ID",
  backQuery = "Back",
  backSelectedImage = "Back ID",
}

export const ExploreDebounceMS = 700;
export const ExplorePageSize = 60;

export const SortByOptions: { [option in SortBy]: string } = {
  dateCreatedDescending: "Date Created (Newest-Oldest)",
  dateCreatedAscending: "Date Created (Oldest-Newest)",
  dateModifiedDescending: "Date Modified (Newest-Oldest)",
  dateModifiedAscending: "Date Modified (Oldest-Newest)",
  nameAscending: "Name (A-Z)",
  nameDescending: "Name (Z-A)",
};
