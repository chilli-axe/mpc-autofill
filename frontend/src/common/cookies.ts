/**
 * Data API for interacting with anything stored in cookies or local storage.
 */

// import Ajv2020 from "ajv/dist/2020";
import Cookies from "js-cookie";

import {
  BackendURLKey,
  CSRFKey,
  GoogleAnalyticsConsentKey,
  SearchSettingsKey,
} from "@/common/constants";
import { Convert } from "@/common/schema_types";
import { SearchSettings, SourceDocuments, SourceRow } from "@/common/types";
import { getSourceRowsFromSourceSettings } from "@/common/utils";
import { getDefaultSearchSettings } from "@/store/slices/searchSettingsSlice";

// import * as SearchSettingsSchema from "../../../schemas/schemas/SearchSettings.json";

// const ajv = new Ajv2020();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFKey);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

/**
 * Get search settings from localStorage data. If valid data is retrieved,
 * ensure that all `sourceDocuments` are included in the returned settings,
 * with any new sources that weren't previously included added to the end and enabled.
 */
export function getLocalStorageSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  const serialisedRawSettings = localStorage.getItem(SearchSettingsKey) ?? "{}";
  try {
    const searchSettings = Convert.toSearchSettings(serialisedRawSettings);
    // great, the user has valid search settings stored in their browser local storage.
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) =>
        parseInt(sourceDocument.pk)
      )
    );
    // types have to be narrowed here because quicktype doesn't support our SourceRow data structure :(
    const sources: Array<SourceRow> = getSourceRowsFromSourceSettings(
      searchSettings.sourceSettings
    );
    const sourceInLocalStorageSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    searchSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInLocalStorageSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return searchSettings;
  } catch (e) {
    // quicktype will throw an error if the user's stored settings do not match the schema
    // e.g. upon first page load
    // just return the default settings
    return getDefaultSearchSettings(sourceDocuments);
  }
}

export function setLocalStorageSearchSettings(settings: SearchSettings): void {
  localStorage.setItem(SearchSettingsKey, JSON.stringify(settings));
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentKey);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentKey, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getLocalStorageBackendURL() {
  return localStorage.getItem(BackendURLKey);
}

export function setLocalStorageBackendURL(url: string) {
  localStorage.setItem(BackendURLKey, url);
}

export function clearLocalStorageBackendURL(): void {
  localStorage.removeItem(BackendURLKey);
}

//# endregion
