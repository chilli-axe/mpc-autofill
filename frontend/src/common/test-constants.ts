/**
 * Some pre-built objects which can be used to build up Redux state for tests.
 */

import { Card, MaximumDPI, MaximumSize, MinimumDPI } from "@/common/constants";
import { CardType as CardTypeSchema, SourceType } from "@/common/schema_types";
import {
  BackendState,
  CardDocument,
  Project,
  SearchSettings,
  SourceDocument,
  SourceDocuments,
} from "@/common/types";

//# region backend

export const localBackendURL = "https://127.0.0.1:8000";
export const localBackend: BackendState = { url: localBackendURL };
export const noBackend: BackendState = { url: null };

//# endregion

//# region sources

export const sourceDocument1: SourceDocument = {
  pk: 0,
  key: "source_1",
  name: "Source 1",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocument2: SourceDocument = {
  pk: 1,
  key: "source_2",
  name: "Source 2",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocument3: SourceDocument = {
  pk: 2,
  key: "source_3",
  name: "Source 3",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocument4: SourceDocument = {
  pk: 3,
  key: "source_4",
  name: "Source 4",
  sourceType: SourceType.GoogleDrive,
  externalLink: undefined,
  description: "",
};

export const sourceDocuments: SourceDocuments = {
  [sourceDocument1.pk]: sourceDocument1,
  [sourceDocument2.pk]: sourceDocument2,
  [sourceDocument3.pk]: sourceDocument3,
  [sourceDocument4.pk]: sourceDocument4,
};

//# endregion

//# region cards

export const cardDocument1: CardDocument = {
  identifier: "abc123",
  cardType: CardTypeSchema.Card,
  name: "Card 1",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 1",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument2: CardDocument = {
  identifier: "abc1234",
  cardType: CardTypeSchema.Card,
  name: "Card 2",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 2",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument3: CardDocument = {
  identifier: "abc12345",
  cardType: CardTypeSchema.Card,
  name: "Card 3",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 3",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument4: CardDocument = {
  identifier: "abc123456",
  cardType: CardTypeSchema.Card,
  name: "Card 4",
  priority: 0,
  source: sourceDocument1.key,
  sourceName: sourceDocument1.name,
  sourceId: sourceDocument1.pk,
  sourceVerbose: sourceDocument1.name,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 4",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument5: CardDocument = {
  identifier: "abc1234567",
  cardType: CardTypeSchema.Cardback,
  name: "Card 5",
  priority: 0,
  source: sourceDocument2.key,
  sourceName: sourceDocument2.name,
  sourceId: sourceDocument2.pk,
  sourceVerbose: `${sourceDocument2.name} Cardbacks`,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 5",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

export const cardDocument6: CardDocument = {
  identifier: "abc12345678",
  cardType: CardTypeSchema.Token,
  name: "Card 6",
  priority: 0,
  source: sourceDocument3.key,
  sourceName: sourceDocument3.name,
  sourceId: sourceDocument3.pk,
  sourceVerbose: `${sourceDocument3.name} Tokens`,
  sourceType: SourceType.GoogleDrive,
  sourceExternalLink: undefined,
  dpi: 1200,
  searchq: "card 6",
  extension: "png",
  dateCreated: "1st January, 2000", // formatted by backend
  dateModified: "1st January, 2000", // formatted by backend
  downloadLink: "",
  size: 10_000_000,
  smallThumbnailUrl: "",
  mediumThumbnailUrl: "",
  language: "EN",
  tags: [],
};

//# endregion

//# region project

export const projectSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectThreeMembersSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectSelectedImage2: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", cardType: Card },
        selectedImage: cardDocument2.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

//# endregion

//# region search settings

export const defaultSettings: SearchSettings = {
  searchTypeSettings: { fuzzySearch: false, filterCardbacks: false },
  sourceSettings: {
    sources: [
      [sourceDocument1.pk, true],
      [sourceDocument2.pk, true],
      [sourceDocument3.pk, true],
      [sourceDocument4.pk, true],
    ],
  },
  filterSettings: {
    minimumDPI: MinimumDPI,
    maximumDPI: MaximumDPI,
    maximumSize: MaximumSize,
    languages: [],
    includesTags: [],
    excludesTags: ["NSFW"],
  },
};

//# endregion
