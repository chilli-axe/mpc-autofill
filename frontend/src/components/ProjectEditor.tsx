/**
 * This component represents the complete MPC Autofill project editor, ready to
 * drop into a page (as the only component). Must be wrapped with a Redux provider.
 */

import React, { useEffect } from "react";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { RibbonHeight } from "@/common/constants";
import { useAppSelector } from "@/common/types";
import { NoBackendDefault } from "@/components/NoBackendDefault";
import { OverflowCol } from "@/components/OverflowCol";
import { Ribbon } from "@/components/Ribbon";
import { SelectedImagesRibbon } from "@/features/bulkManagement/SelectedImagesRibbon";
import { CardGrid } from "@/features/card/CardGrid";
import { CommonCardback } from "@/features/card/CommonCardback";
import { Export } from "@/features/export/Export";
import { FinishedMyProject } from "@/features/export/FinishedMyProjectModal";
import { FinishSettings } from "@/features/finishSettings/FinishSettings";
import { Import } from "@/features/import/Import";
import { SearchSettings } from "@/features/searchSettings/SearchSettings";
import { Status } from "@/features/status/Status";
import { useBackendConfigured } from "@/store/slices/backendSlice";
import {
  selectIsProjectEmpty,
  selectProjectCardback,
} from "@/store/slices/projectSlice";

function ProjectEditor() {
  // TODO: should we periodically ping the backend to make sure it's still alive?
  //# region queries and hooks

  const backendConfigured = useBackendConfigured();
  const cardback = useAppSelector(selectProjectCardback);
  const isProjectEmpty = useAppSelector(selectIsProjectEmpty);

  //# endregion

  //# region effects

  /**
   * Ask the user for confirmation before they close the page if their project has any cards in it.
   */
  useEffect(() => {
    const handler = (event: BeforeUnloadEvent) => {
      if (!isProjectEmpty) {
        event.preventDefault();
        return false;
      }
    };
    window.addEventListener("beforeunload", handler);
    return () => {
      window.removeEventListener("beforeunload", handler);
    };
  }, [isProjectEmpty]);

  //# endregion

  return (
    <>
      {backendConfigured ? (
        <Row className="g-0">
          <Ribbon className="g-0">
            <SelectedImagesRibbon />
          </Ribbon>
          <OverflowCol
            lg={8}
            md={8}
            sm={6}
            xs={6}
            data-testid="left-panel"
            heightDelta={RibbonHeight}
          >
            <CardGrid />
          </OverflowCol>
          <OverflowCol
            data-testid="right-panel"
            lg={4}
            md={4}
            sm={6}
            xs={6}
            style={{ zIndex: 1 }}
            className="px-2"
            heightDelta={RibbonHeight}
          >
            <Status />
            <Row className="g-0 pt-2">
              <FinishSettings />
            </Row>
            <Row className="g-0 pt-2">
              <SearchSettings />
            </Row>
            <Row className="g-0 pt-2">
              <Col lg={7} md={12} sm={12} xs={12}>
                <Import />
              </Col>
              <Col lg={5} md={12} sm={12} xs={12}>
                <Export />
              </Col>
              <FinishedMyProject />
            </Row>
            <Col className="g-0 pt-2" lg={{ span: 8, offset: 2 }} md={12}>
              <CommonCardback selectedImage={cardback} />
            </Col>
          </OverflowCol>
        </Row>
      ) : (
        <NoBackendDefault />
      )}
    </>
  );
}

export default ProjectEditor;
