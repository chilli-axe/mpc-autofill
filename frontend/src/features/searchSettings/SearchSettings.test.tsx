import { waitFor, within } from "@testing-library/dom";

import { Front } from "@/common/constants";
import { cardDocument1, sourceDocument1 } from "@/common/test-constants";
import {
  expectCardGridSlotState,
  importText,
  openSearchSettingsModal,
  renderWithProviders,
} from "@/common/test-utils";
import ProjectEditor from "@/components/ProjectEditor";
import {
  cardbacksOneResult,
  cardDocumentsOneResult,
  defaultHandlers,
  searchResultsOneResult,
  sourceDocumentsThreeResults,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

//# region snapshot tests

test("the html structure of search settings", async () => {
  server.use(
    cardDocumentsOneResult,
    cardbacksOneResult,
    sourceDocumentsThreeResults,
    searchResultsOneResult,
    ...defaultHandlers
  );
  const { container } = renderWithProviders(<ProjectEditor />, {
    preloadedState: {
      project: { members: [], cardback: null, mostRecentlySelectedSlot: null },
    },
  });

  // we need to wait for a second here after rendering to ensure sources have been fetched
  // searching for a card here isn't strictly necessary for the snapshot test,
  // but it's a convenient way to ensure that sources have been fetched
  await importText("my search query");
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);

  const searchSettings = await openSearchSettingsModal();
  await waitFor(() => within(searchSettings).getByText(sourceDocument1.name));
  // the below line waits until all multi-select elements have finished loading
  await waitFor(() => container.querySelectorAll(".spinner").length === 0);
  expect(searchSettings).toMatchSnapshot();
});

//# endregion
