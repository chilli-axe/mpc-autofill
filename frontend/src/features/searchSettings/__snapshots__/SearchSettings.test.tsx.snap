// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of search settings 1`] = `
.c0 {
  color: black;
}

.c0 .tag {
  color: black;
  background-color: #dddddd;
}

.c0 .tag-remove {
  color: #666666;
}

.c0 .dropdown-trigger {
  border-radius: 0.25rem;
  background-color: white;
}

.c0 .dropdown-content {
  border-radius: 0.25rem;
}

.c0 .search {
  background-color: white;
  color: black;
}

.c0 .search::-webkit-input-placeholder {
  color: black;
}

.c0 .search::-moz-placeholder {
  color: black;
}

.c0 .search:-ms-input-placeholder {
  color: black;
}

.c0 .search::placeholder {
  color: black;
}

.c0 .placeholder {
  color: gray;
}

.c0 .toggle {
  font: normal normal normal 12px/1 bootstrap-icons;
  top: 2px;
  left: 2px;
}

.c0 .toggle.collapsed::after {
  content: "\\F4FA";
}

.c0 .toggle.expanded::after {
  content: "\\F2E6";
}

.c0 .root {
  padding: 0;
  margin: 0;
}

.c1 {
  font-size: 1em;
  cursor: pointer;
}

<div
  class="modal-dialog modal-dialog-scrollable"
  data-testid="search-settings"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Search Settings
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <div
        class="px-1 container"
      >
        <h5>
          Search Type
        </h5>
        Configure how closely the search results should match your query.
        <br />
        <br />
        <div
          class="btn toggle btn-md off btn-info"
          role="button"
          style="width: 100%; height: 38px;"
        >
          <div
            class="toggle-group"
          >
            <span
              class="btn toggle-on btn-md flex-centre btn-success"
            >
              Fuzzy (Forgiving) Search
            </span>
            <span
              class="btn toggle-off btn-md flex-centre btn-info"
            >
              Precise Search
            </span>
            <span
              class="toggle-handle btn btn-md btn-default"
            />
          </div>
        </div>
        <br />
        <br />
        <div
          class="btn toggle btn-md off btn-info"
          role="button"
          style="width: 100%; height: 38px;"
        >
          <div
            class="toggle-group"
          >
            <span
              class="btn toggle-on btn-md flex-centre btn-success"
            >
              Filters Apply to Cardbacks
            </span>
            <span
              class="btn toggle-off btn-md flex-centre btn-info"
            >
              Include All Cardbacks
            </span>
            <span
              class="toggle-handle btn btn-md btn-default"
            />
          </div>
        </div>
      </div>
      <hr />
      <div
        class="px-1 container"
      >
        <h5>
          Filters
        </h5>
        Configure the DPI (dots per inch) and file size ranges the search results must be within.
        <br />
        At a fixed physical size, a higher DPI yields a higher resolution print.
        <br />
        MakePlayingCards prints cards up to 
        <b>
          800 DPI
        </b>
        , meaning an 800 DPI print and a 1200 DPI print will 
        <b>
          look the same
        </b>
        .
        <br />
        <br />
        <div
          class="row"
        >
          <div
            class="col-6"
          >
            <label
              class="form-label"
            >
              Minimum: 
              <b>
                0
                 DPI
              </b>
            </label>
            <input
              class="form-range"
              max="1500"
              min="0"
              step="50"
              type="range"
              value="0"
            />
          </div>
          <div
            class="col-6"
          >
            <label
              class="form-label"
            >
              Maximum: 
              <b>
                1500
                 DPI
              </b>
            </label>
            <input
              class="form-range"
              max="1500"
              min="0"
              step="50"
              type="range"
              value="1500"
            />
          </div>
        </div>
        <label
          class="form-label"
        >
          File size: Up to 
          <b>
            30
             MB
          </b>
        </label>
        <input
          class="form-range"
          max="30"
          min="0"
          step="1"
          type="range"
          value="30"
        />
        <br />
        <br />
        Configure the languages and tags to filter the search results on.
        <br />
        <br />
        <label
          class="form-label"
          for="selectLanguage"
        >
          Select languages
        </label>
        <div
          class="c0 react-dropdown-tree-select"
          id="rdts1"
        >
          <div
            class="dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="tree"
              class="dropdown-trigger arrow bottom"
              id="rdts1_trigger"
              role="button"
              tabindex="0"
            >
              <ul
                class="tag-list"
              >
                <li
                  class="tag-item"
                >
                  <span
                    class="placeholder"
                  >
                    Choose...
                  </span>
                </li>
              </ul>
            </a>
          </div>
        </div>
        <label
          class="form-label"
          for="selectTags"
        >
          Select tags which cards must have 
          <b>
            at least one
          </b>
           of
        </label>
        <div
          class="c0 react-dropdown-tree-select"
          id="rdts2"
        >
          <div
            class="dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="tree"
              class="dropdown-trigger arrow bottom"
              id="rdts2_trigger"
              role="button"
              tabindex="0"
            >
              <ul
                class="tag-list"
              >
                <li
                  class="tag-item"
                >
                  <span
                    class="placeholder"
                  >
                    Choose...
                  </span>
                </li>
              </ul>
            </a>
          </div>
        </div>
        <label
          class="form-label"
          for="selectTags"
        >
          Select tags which cards must 
          <b>
            not
          </b>
           have
        </label>
        <div
          class="c0 react-dropdown-tree-select"
          id="rdts3"
        >
          <div
            class="dropdown"
          >
            <a
              aria-expanded="false"
              aria-haspopup="tree"
              class="dropdown-trigger arrow bottom"
              id="rdts3_trigger"
              role="button"
              tabindex="0"
            >
              <ul
                class="tag-list"
              >
                <li
                  class="tag-item"
                >
                  <span
                    class="placeholder"
                  >
                    Choose...
                  </span>
                </li>
              </ul>
            </a>
          </div>
        </div>
      </div>
      <hr />
      <div
        class="px-1 container"
      >
        <h5>
          Sources
        </h5>
        Configure the sources to include in the search results.
        <ul>
          <li>
            <b>
              Drag & drop
            </b>
             them to change the order they're searched in.
          </li>
          <li>
            Use the 
            <b>
              arrows
            </b>
             to send a source to the top or bottom.
          </li>
        </ul>
        <div
          class="d-grid gap-0 mt-3"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Disable
             all drives
          </button>
        </div>
        <br />
        <table
          class="table"
          style="table-layout: auto;"
        >
          <thead>
            <tr
              style="height: 38px;"
            >
              <th
                class="prevent-select"
              >
                Enabled
              </th>
              <th
                class="prevent-select"
              >
                Source Name
              </th>
              <th />
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
              data-rfd-drag-handle-context-id=":r0:"
              data-rfd-drag-handle-draggable-id="0"
              data-rfd-draggable-context-id=":r0:"
              data-rfd-draggable-id="0"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                style="vertical-align: middle; width: 30%;"
              >
                <div
                  class="btn toggle btn-md btn-primary"
                  role="button"
                  style="width: 80%; height: 38px;"
                >
                  <div
                    class="toggle-group"
                  >
                    <span
                      class="btn toggle-on btn-md flex-centre prevent-select btn-primary"
                    >
                      On
                    </span>
                    <span
                      class="btn toggle-off btn-md flex-centre prevent-select btn-secondary"
                    >
                      Off
                    </span>
                    <span
                      class="toggle-handle btn btn-md btn-default"
                    />
                  </div>
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 50%;"
              >
                Source 1
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <div>
                  <i
                    class="c1 bi bi-chevron-double-up"
                  />
                </div>
                <div>
                  <i
                    class="c1 bi bi-chevron-double-down"
                  />
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 15%; text-align: center;"
              >
                <i
                  class="bi bi-grip-horizontal"
                  style="font-size: 2em;"
                />
              </td>
            </tr>
            <tr
              aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
              data-rfd-drag-handle-context-id=":r0:"
              data-rfd-drag-handle-draggable-id="1"
              data-rfd-draggable-context-id=":r0:"
              data-rfd-draggable-id="1"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                style="vertical-align: middle; width: 30%;"
              >
                <div
                  class="btn toggle btn-md btn-primary"
                  role="button"
                  style="width: 80%; height: 38px;"
                >
                  <div
                    class="toggle-group"
                  >
                    <span
                      class="btn toggle-on btn-md flex-centre prevent-select btn-primary"
                    >
                      On
                    </span>
                    <span
                      class="btn toggle-off btn-md flex-centre prevent-select btn-secondary"
                    >
                      Off
                    </span>
                    <span
                      class="toggle-handle btn btn-md btn-default"
                    />
                  </div>
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 50%;"
              >
                Source 2
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <div>
                  <i
                    class="c1 bi bi-chevron-double-up"
                  />
                </div>
                <div>
                  <i
                    class="c1 bi bi-chevron-double-down"
                  />
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 15%; text-align: center;"
              >
                <i
                  class="bi bi-grip-horizontal"
                  style="font-size: 2em;"
                />
              </td>
            </tr>
            <tr
              aria-describedby="rfd-hidden-text-:r0:-hidden-text-:r1:"
              data-rfd-drag-handle-context-id=":r0:"
              data-rfd-drag-handle-draggable-id="2"
              data-rfd-draggable-context-id=":r0:"
              data-rfd-draggable-id="2"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <td
                style="vertical-align: middle; width: 30%;"
              >
                <div
                  class="btn toggle btn-md btn-primary"
                  role="button"
                  style="width: 80%; height: 38px;"
                >
                  <div
                    class="toggle-group"
                  >
                    <span
                      class="btn toggle-on btn-md flex-centre prevent-select btn-primary"
                    >
                      On
                    </span>
                    <span
                      class="btn toggle-off btn-md flex-centre prevent-select btn-secondary"
                    >
                      Off
                    </span>
                    <span
                      class="toggle-handle btn btn-md btn-default"
                    />
                  </div>
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 50%;"
              >
                Source 3
              </td>
              <td
                style="vertical-align: middle; width: 5%; text-align: center;"
              >
                <div>
                  <i
                    class="c1 bi bi-chevron-double-up"
                  />
                </div>
                <div>
                  <i
                    class="c1 bi bi-chevron-double-down"
                  />
                </div>
              </td>
              <td
                style="vertical-align: middle; width: 15%; text-align: center;"
              >
                <i
                  class="bi bi-grip-horizontal"
                  style="font-size: 2em;"
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close Without Saving
      </button>
      <button
        class="btn btn-primary"
        type="button"
      >
        Save Changes
      </button>
    </div>
  </div>
</div>
`;
